projects/plugins/jetpack/extensions/plugins/ai-assistant-plugin/components/breve/utils/get-feature-data.ts

```diff
@@ -0,0 +1,173 @@
+/**
+ * External dependencies
+ */
+import { dispatch } from '@wordpress/data';
+import debugFactory from 'debug';
+/**
+ * Types
+ */
+import { BreveDispatch, SpellingDictionaryContext } from '../types';
+
+const debug = debugFactory( 'jetpack-ai-breve:fetch-feature-data' );
+
+type GetFeatureDataParams = {
+	feature: string;
+	language: string;
+	lastRequested?: string;
+};
+
+const briefData: {
+	// feature
+	[ key: string ]: {
+		// language
+		[ key: string ]: {
+			loading: boolean;
+			failed: boolean;
+			verified?: boolean;
+			data?: Array< string > | SpellingDictionaryContext;
+		};
+	};
+} = {};
+
+// Clear the local storage for the spelling context's old key. TODO: remove this after a few releases
+localStorage.removeItem( 'jetpack-ai-breve-spelling-context-en' );
+
+async function fetchFromServer( {
+	feature,
+	language,
+	lastRequested,
+}: GetFeatureDataParams ): Promise< {
+	requestTime: string;
+	data: Array< string > | SpellingDictionaryContext | null;
+} > {
+	// Randomize the server to balance the load
+	const counter = Math.floor( Math.random() * 3 );
+	const url = `https://s${ counter }.wp.com/wp-content/lib/jetpack-ai/breve-dictionaries/${ feature }/${ language }.json`;
+
+	// If we have a lastRequested date, first send a HEAD request to check if the data has been modified
+	// The If-Modified-Since header causes a CORS preflight request error, so we need to check manually
+	if ( lastRequested ) {
+		const headData = await fetch( url, { method: 'HEAD' } );
+		const lastModified = headData.headers.get( 'last-modified' );
```

Comment:
The `etag` should be more precise, but it is always `null` for some reason, even though I can see it in the browser tools :thinking: